src/main/java/cubicchunks/util/WatchersSortingList.java

```diff
@@ -0,0 +1,165 @@
+
+    /** Sort elements */
+    public void sort() {
+        Arrays.sort((T[])data, 0, size, order);
+    }
+
+    /** Check if list is empty. Even when it return true list still could contain 
+     * data, but it will be removed in attempt to access it via iterator.
+     * @return if list contain any accessible data */
+    public boolean isEmpty() {
+        return size - toRemove.size() == 0;
+    }
+
+    /** Return iterator over elements of list. If iterator meet any element which is scheduled to be removed
+     * by {@code remove(T element)} function, it will be removed from data.
+     * @return iterator over elements */
+    public Iterator<T> iterator() {
+        return new Iterator<T>() {
+
+            int a = 0;
+            T prev = null;
+            T next = null;
+
+            private void peekNext() {
+                while (next == null && a < size) {
+                    T e = (T) data[++a - 1];
+                    if (toRemove.remove(e)) {
+                        dataAsSet.remove(e);
+                        data[--a] = data[--size];
+                    } else {
+                        next = e;
+                    }
+                }
+            }
+
+            @Override
+            public boolean hasNext() {
+                peekNext();
+                return next != null;
+            }
+
+            @Override
+            public T next() {
+                peekNext();
+                if (next == null)
+                    throw new ArrayIndexOutOfBoundsException();
+                prev = next;
+                next = null;
+                return prev;
+            }
+
+            @Override
+            public void remove() {
+                dataAsSet.remove(data[--a]);
+                data[a] = data[--size];
+            }
+        };
+    }
+    /** Schedule element to be removed. Function is not save, you need to check 
+     * explicitly if this list contain element which need to be removed.
+     * Attempt to remove non-existing elements will cause nasty bugs.*/
+    public void remove(T entry) {
+        toRemove.add(entry);
+    }
+
+    /** Schedule such elements {@code a} whom return {@code true} on call {@code predicate.test(a)} 
+     * to be removed from that list during iteration. */
+    public void removeIf(Predicate<T> predicate) {
+        for (int i = 0; i < size; i++) {
+            T a = (T) data[i];
+            if (predicate.test(a))
+                toRemove.add(a);
+        }
+    }
+
+    /** Add element T at desired position in list.
+     * @throws  ArrayIndexOutOfBoundsException if index is less than a 0 or bigger than index of a last element. */
+    public void add(int index, T element) {
+        if (index < 0 || index >= size)
+            throw new ArrayIndexOutOfBoundsException();
+        if (size == 0) {
+            this.add(element);
+            return;
+        }
+        Object old = data[index];
+        data[index] = element;
+        this.add((T) old);
```

Comment:
This function used here: https://github.com/OpenCubicChunks/CubicChunks/blob/753ef1ca306c30989e0a24e055c7ca36ab2cd1ab/src/main/java/cubicchunks/server/PlayerCubeMap.java#L370
To add cube to generation query with high priority.

Reply:
And your code completely missed the point as when a second none is added the first one will be moved. To the end. Just make the values stored internally in reverse order so you can have insertFirst method and it would only need to append it at the end of the array.